# Tic-Tac-Toe (Extra Credit Assignment)

##### Author: Yuji Mori
##### Submission Date: 01/20/2021
##### STATS404 Winter 2021

In [1]:
import random
random.seed(1234)

In [24]:
def write_to_board(board_state,key,mark):
    '''insert value (X or O) into board at the given key position'''
    board_state[key] = mark

In [37]:
def check_for_winner(board_state,turn):
    ''' Determine if the game has a winner at the current board state.
        The logic is organized by focusing on three key 'pivot' points on the board 
        (there is probably a much simpler way that doesn't rely on hard-coding logic)
    Args:
        board_state (dict): dictionary with key/values representing state of tic-tac-toe board
        turn (int): index for turn count
    Returns:
        winner (str): If winner is found, returns string 'X' or 'O'. 
                     'Tie' if tie on turn 9. 
                     'None' if game is still in progress.
    '''
    # check victories that require top-L 
    if (board['top-L'] != ' ' 
        and (
            # horizonal win
            (board['top-L'] == board['top-M'] == board['top-R']) or
            # vertical win
            (board['top-L'] == board['mid-L'] == board['low-L'])
        )
    ): 
        winner = board['top-L']
    # check victories that require mid-M
    elif (board['mid-M'] != ' ' 
        and (
            # diagonal win
            (board['top-L'] == board['mid-M'] == board['low-R']) or
            (board['low-L'] == board['mid-M'] == board['top-R']) or
            # horizonal win
            (board['mid-L'] == board['mid-M'] == board['mid-R']) or
            # vertical win
            (board['top-M'] == board['mid-M'] == board['low-M']) 
        )
    ): 
        winner = board['mid-M']
    # check victories that require low-R
    elif (board['low-R'] != ' ' 
        and (
            # horizonal win
            (board['low-L'] == board['low-M'] == board['low-R']) or
            # vertical win
            (board['top-R'] == board['mid-R'] == board['low-R'])
        )
    ): 
        winner = board['low-R']
        
    # declare tie if board is full
    elif (turn == 9):
        winner = 'Tie'
        
    # otherwise, continue game
    else:
        winner = 'None'
        
    return(winner)


In [47]:
# create list to keep track of victories over 50,000 iterations
winners_list = []

for i in range(50000):
    
    # initialize empty board
    board = {'top-L': ' ', 'top-M': ' ', 'top-R': ' ',
             'mid-L': ' ', 'mid-M': ' ', 'mid-R': ' ',
             'low-L': ' ', 'low-M': ' ', 'low-R': ' '}

    # creating a list of available board positions, remove elements as turns are taken
    empty_spots = list(board.keys())

    # initialize game state
    game_over = False

    # initialize starting player
    x_or_o = ['X','O']
    current_player = 0
    turn = 1

    # begin game
    while (game_over == False):
        # pick a random empty position and make move
        index = random.randint(0,len(empty_spots)-1)
        write_to_board(board,empty_spots[index], x_or_o[current_player])
        # check for victory
        game_winner = check_for_winner(board,turn)
        if (game_winner != 'None'): 
            game_over = True 
            # record winner in list
            winners_list.append(game_winner)
        else:
            # update list of available spots
            empty_spots.pop(index)
            # switch players
            current_player = abs(current_player - 1)
            # increment turn counter
            turn = turn + 1 


In [48]:
print(winners_list.count('X'))
print(winners_list.count('X')/len(winners_list))

29206
0.58412


In [49]:
print(winners_list.count('O'))
print(winners_list.count('O')/len(winners_list))

14489
0.28978


In [50]:
print(winners_list.count('Tie'))
print(winners_list.count('Tie')/len(winners_list))

6305
0.1261


### Discussion

After 50,000 iterations of the tic-tac-toe game, ~58% of the games were 'X' win, ~29% were 'O' win, and ~13% were ties.

I found that player 'X' has a much higher win rate than player 'O'. The only difference is the turn order ('X' always goes first), therefore we can postulate that the 1st move gives a player a significant advantage **when moves are randomized.** 

In a real game with human players, the these win rates may be very different. While a first-turn advantange may still exist, each player will actively pursue and deny 3-in-a-row victories (unlike this program, where the moves are made without any regard for the win condition). I would expect the majority of outcomes to be ties in a real-world scenario, and win rates between 'X' and 'O' should be much closer.